### PR TITLE
Iz gawb 1743 modal for importing data from another workspace

### PR DESCRIPTION
**Pull request by zarsky-broad**
(potentially a few text formatting changes incoming, but this is the major parts, anyway)

There's a mock checked in for testing soft conflicts.

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/cljs/main/broadfcui/common/components.cljs [hunk 1]
```diff
@@ -273,38 +273,52 @@
...
+               ;; Padding is specifically em rather than rem to match fa-fw
+               [:div {:style {:overflowX "auto" :paddingLeft "1.3em"}}
+                [:div {} (str "Code: " status-code)]
+                (when timestamp [:div {} "Occurred: "
+                                 (common/format-date timestamp
+                                                     (assoc common/default-date-format
+                                                       :timeZoneName "short"))])
+                (when source [:div {} "Source: " source])
+                (when (seq causes)
+                  (let [num-hidden (- (count causes) 4)]
```
Comment (dmohs) [inline src/cljs/main/broadfcui/common/components.cljs]:
```cljs
(let [shown-cause-count 4
      hidden-count (- (count causes) shown-cause-count)
...
```

File: src/cljs/main/broadfcui/common/components.cljs [hunk 2]
```diff
@@ -675,3 +689,50 @@
    (fn [{:keys [refs]}]
      (.select2 (js/$ (@refs "input-element")) "destroy"))})
 
+
```
Comment (dmohs) [inline src/cljs/main/broadfcui/common/components.cljs]:
```cljs
;; This component is called recursively.
```

File: src/cljs/main/broadfcui/common/components.cljs [hunk 3]
```diff
@@ -675,3 +689,50 @@
...
+(declare Tree)
+
```
Comment (dmohs) [inline src/cljs/main/broadfcui/common/components.cljs]: If you had this:
```cljs
(defn- map-node [node f]
  (map
    (fn [key]
      (let [value (get node key)
          (f key value (not (and (vector? value) (not-empty value))))))
  (keys node))
```

File: src/cljs/main/broadfcui/common/components.cljs [hunk 4]
```diff
@@ -675,3 +689,50 @@
...
+     {:collapsed? (or (:start-collapsed? props) false)})
+   :render
+   (fn [{:keys [props state]}]
+     (let [body
+           [:div {:hidden (and (:collapsed? @state) (:label props))
+                  :marginLeft (if (:label props) "0.5rem" 0)}
+            (map (fn [node]
+                   [:ul {:style {:margin "0.2rem" :padding "0.5rem"
+                                 :backgroundColor (if (and (:highlight-ends? props)
+                                                           (not-any? (fn [key]
```
Comment (dmohs) [inline src/cljs/main/broadfcui/common/components.cljs]: you could then do:
```cljs
(every? identity (map-node node (fn [_ _ leaf?] leaf?)))
```

File: src/cljs/main/broadfcui/common/components.cljs [hunk 5]
```diff
@@ -675,3 +689,50 @@
...
+                                                                       (let [value (get node key)]
+                                                                         (and (vector? value)
+                                                                              (not-empty value))))
+                                                                     (keys node)))
+                                                    (:warning-state style/colors)
+                                                    "rgba(0,0,0,0.1)")
+                                 :borderRadius 8}}
+                    (map (fn [key]
```
Comment (dmohs) [inline src/cljs/main/broadfcui/common/components.cljs]: and:
```cljs
(map-node
  node
  (fn [key value leaf?]
    [:li {:style {:listStyle "none" :padding "0.1rem"}}
    (if leaf?
      [:span {} [:strong {} key ": "] value]
      [Tree {:data value
             :start-collapsed? (:start-collapsed? props)
             :highlight-ends? (:highlight-ends? props)
             :label [:strong {} key ":"]}]
...
```

File: src/cljs/main/broadfcui/endpoints.cljs [hunk 6]
```diff
@@ -626,8 +626,9 @@
    :method :post})
 
 
-(defn copy-entity-to-workspace [workspace-id]
-  {:path (str "/workspaces/" (ws-path workspace-id) "/entities/copy")
+(defn copy-entity-to-workspace [workspace-id re-link-soft-conflicts?]
+  {:path (str "/workspaces/" (ws-path workspace-id)
+              "/entities/copy?linkExistingEntities=" (true? re-link-soft-conflicts?))
```
Comment (dmohs) [inline src/cljs/main/broadfcui/endpoints.cljs]:
```cljs
(boolean re-link-soft-conflicts?)
```
since a user of this function might want to pass any non-false value to trigger this flag.

File: src/cljs/main/broadfcui/page/workspace/data/copy_data_entities.cljs [hunk 7]
```diff
@@ -18,37 +22,112 @@
...
+                      [:p {} "The conflicting entities are highlighted below."]
+                      [comps/Tree {:highlight-ends? true
+                                   :data soft-conflicts}]
+                      [:p {} "You may link the " import-type "s that you import to the existing
+                      entities in the destination workspace, by clicking " [:strong {} "Re-link"] "."]
+                      [:p {} "Re-linking will not import the conflicting entities from, or change the "
+                       import-type "s in, the source workspace."
+                       (common/render-info-box
+                        {:position "top"
+                         :text (react/create-element
```
Comment (dmohs) [inline src/cljs/main/broadfcui/page/workspace/data/copy_data_entities.cljs]: We generally use `create-element` only when we need to supply refs and need those to be associated with our component. Since you're not supplying refs here, you should be able to omit this (and the info box component should support that).

File: src/styles/foundation.scss [hunk 8]
```diff
@@ -33,4 +33,5 @@ $tooltip-font-size: 0.8rem;
 
 .dropdown-pane {
   font-weight: normal;
+  z-index: 523;
```
Comment (dmohs) [inline src/styles/foundation.scss]: comment on why this value

File: src/cljs/main/broadfcui/page/workspace/summary/attribute_editor.cljs [hunk 9]
```diff
@@ -106,92 +106,92 @@
...
-                                                         {:id "focus"}))))}
-                         {:header (header "Value") :starting-width :remaining :as-text (constantly nil) :resizable? false
-                          :content-renderer
-                          (fn [{:keys [value index]}]
-                            (style/create-text-field {:style {:marginBottom 0 :fontSize "100%" :height 26 :width "calc(100% - 2px)"}
-                                                      :defaultValue value
+         [:div {}
+          (if editing?
+            [:div {:style {:marginBottom "0.25em"}}
+             [comps/Button {:icon :add-new :text "Add new"
```
Comment (dmohs) [inline src/cljs/main/broadfcui/page/workspace/summary/attribute_editor.cljs]: This is actually the only change in this file. If you want to reformat, please make a separate PR for that.

**Review by dmohs — COMMENTED** [review]
Make changes and then I'll run this and test a bit. Code looks great.

**Review by dmohs — APPROVED** [review]
One suggestion. Looks great.

File: src/cljs/main/broadfcui/page/workspace/data/copy_data_entities.cljs [hunk 10]
```diff
@@ -18,37 +22,111 @@
...
+                  (react/call :show-import-result this (get-parsed-response false) selected))}))
+   :show-import-result
+   (fn [{:keys [this props]} parsed-response selected]
+     (let [formatted-response (postwalk
+                               (fn [x]
+                                 (cond
+                                   (= x "entityName") "ID"
+                                   (= x "entityType") "Type"
+                                   (= x "conflicts") "Conflicting linked entities"
+                                   :else x))
```
Comment (MattPutnam) [inline src/cljs/main/broadfcui/page/workspace/data/copy_data_entities.cljs]: You should be able to do this with a `case` instead.

File: src/cljs/main/broadfcui/page/workspace/summary/attribute_editor.cljs [hunk 11]
```diff
@@ -106,92 +106,92 @@
...
+                        {:header (header "Value") :starting-width :remaining :as-text process-attribute-value
+                         :content-renderer (comp (table-utils/render-gcs-links (:workspace-bucket props)) process-attribute-value)}])
+            :data (if editing?
+                    (map-indexed (fn [index [key value type]]
+                                   {:index index :key key :value value :type type})
+                                 (:attributes @state))
+                    (:workspace-attributes props))
+            :->row (if editing?
+                     (juxt :index identity identity identity)
+                     identity)}]])]))
```
Comment (MattPutnam) [inline src/cljs/main/broadfcui/page/workspace/summary/attribute_editor.cljs:194]: Not your code, just commenting: This expression is ridiculous.

**Review by MattPutnam — APPROVED** [review]
